core/src/main/scala/org/apache/spark/sql/delta/files/TahoeCDCBaseFileIndex.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.files
+
+
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.delta.actions.SingleAction.addFileEncoder
+import org.apache.spark.sql.delta.actions.{AddCDCFile, AddFile, FileAction, RemoveFile}
+import org.apache.spark.sql.delta.commands.cdc.CDCReader.{CDCDataSpec, CDC_COMMIT_TIMESTAMP, CDC_COMMIT_VERSION}
+import org.apache.spark.sql.delta.{DeltaLog, Snapshot}
+
+/**
+ * A base [[TahoeFileIndex]] for all CDC file indexes
+ */
+abstract class TahoeCDCBaseFileIndex[T <: FileAction](
+    spark: SparkSession,
+    val filesByVersion: Seq[CDCDataSpec[T]],
+    deltaLog: DeltaLog,
+    path: Path,
+    snapshot: Snapshot) extends TahoeFileIndex(spark, deltaLog, path) {
+
+  case class ActionParameters(
```

Comment:
do we really need this extraction stuff? since T can only be FileAction.. FileAction base trait always has `partitionValue`, `size` (as `getSize`), `dataChange` and `tags` available. The only place where we need something customer is modification time, which is 0 for everything other than AddFile, which can be inline in this base `TahoeCDCBaseFileIndex`. Also, there is some custom check for `RemoveFile` having extended metadata which can also be inlined.